client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,77 +20,83 @@ import {
-			auth_type: authType,
-			two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
-			remember_me: getRememberMe( store.getState() ),
-			two_step_push_token: getTwoFactorPushToken( store.getState() ),
-			client_id: config( 'wpcom_signup_id' ),
-			client_secret: config( 'wpcom_signup_key' ),
-		} ).then( () => {
-			store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
-		} ).catch( error => {
-			store.dispatch( {
-				type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
-				nonceType: 'push',
-				twoStepNonce: error.response.body.data.two_step_nonce
-			} );
-			return Promise.reject( error );
-		} );
+	store.dispatch(
+		http( {
+			url: 'https://wordpress.com/wp-login.php?action=two-step-authentication-endpoint',
+			method: 'POST',
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
+	}
+
+	store.dispatch( {
+		type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
+		nonceType: 'push',
+		twoStepNonce,
+	} );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
```

Comment:
> could you maybe make a simple pseudocode example?

sure! I wrote up one of these kinds of polling mechanisms but ended up abandoning it for unrelated issues. the [README](https://github.com/Automattic/wp-calypso/blob/039bd69fb05835d49e730be5f18be519f7773bcc/client/state/data-layer/wpcom/sites/automated-transfer/status/README.md) describes the role that the data layer handler plays while [the code](https://github.com/Automattic/wp-calypso/pull/10544/files#diff-5c829e44ce68808c631c766cd14577a1R134) in the handler does the actual dispatch/execution

essentially the idea is that the UI just says, "hey, I need to know the status of my 2FA request" and it performs its own logic based on the status info in state. for this example let's suppose that the follow-up response for a successful 2FA exchange is routing to a new path in Calypso

```js
// login form JSX
return (
	<div>
		<Query2FAStatus />
		{ ! state.2fa && showAuthForm() }
		{ state.2fa.error && showAuthFailureForm() }
	</div>
);
```

here the UI only cares about the current status of the attempt. it's either failed or hasn't been attempted (or maybe too old or something like that which is meaningful to the attempt).

```js
// data layer handler
const queueStatusPoll = ( { dispatch }, action ) => {
	dispatch( http( { … } ) ) // make the status poll
}

const interpretSuccess = ( { dispatch }, action, next, data ) => {
	if ( data.notDoneYet ) {
		setTimeout( () => dispatch( action ), POLL_DELAY )
	}

	// was a success
	page( action.followUpPath ) // or something like `/me/settings`
}

const requeue = ( { dispatch }, action, next, error ) => {
	const retryCount = get( action, 'meta.retryCount', 0 );

	if ( retryCount > MAX_RETRY_ATTEMPTS ) {
		return dispatch( createErrorNotice( 'We tried to authenticate but could not reach the server. Please try again later.' ) )
	}

	setTimeout( () => dispatch( merge( {}, action, { meta: { retryCount: retryCount + 1 } } ) );
}

export default {
	[ 2FA_ATTEMPT_REQUEST ]: [ dispatchRequest( queueStatusPoll, interpretSuccess, requeue ) ]
}
```

Here the data layer takes care of all polling and additionally it stores any polling-necessary state _in the action itself_. This is convenient for testing (thanks @samouri for pushing me on this point) so that we can say, "it should give up with an error notice after failing for the third time…"

```js
const action = merge( {}, twoFactorAuth, { meta: { retryCount: 3 } } )
handleTwoFactorAuth( store, action )
expect( dispatch ).to.have.been.calledWithMatch( { type: NOTICE, noticeType: 'error' } )
```

hope this makes sense. I find that it ends up simpler where the UI is happier not knowing the polling details, we have fewer action types to think about, and we can more easily change the behaviors since all the logic is contained here. would love to see some data-layer polling as my linked PR is the only one I know of which takes advantage of these opportunities, but it never merged.